contrib/cell_star/utils/image_util.py

```diff
@@ -0,0 +1,402 @@
+    ys, xs = (lp[0], hp[0]), (lp[1], hp[1])
+    morphology_element = get_circle_kernel(radius)
+    dilated_part = sp.ndimage.morphology.binary_dilation(image[ys[0]:ys[1], xs[0]:xs[1]], morphology_element)
+    image[ys[0]:ys[1], xs[0]:xs[1]] = dilated_part
+    return image
+
+
+def image_dilate_with_element(image, n):
+    return sp.ndimage.morphology.grey_dilation(image, size=(n, n))
+
+
+def image_erode(image, radius):
+    morphology_element = get_circle_kernel(radius)
+    return sp.ndimage.morphology.binary_erosion(image, morphology_element)
+
+
+def fill_foreground_holes(mask, kernel_size, minimal_hole_size, min_cluster_area_scaled, mask_min_radius_scaled):
+    filled_black_holes = fill_holes(mask, kernel_size, minimal_hole_size)
+
+    holes_remaining = np.logical_not(filled_black_holes)
+    filled_small_holes = mark_small_areas(holes_remaining, min_cluster_area_scaled, filled_black_holes)
+
+    morphology_enhanced = image_erode(filled_small_holes, mask_min_radius_scaled)
+    morphology_enhanced = image_dilate(morphology_enhanced, mask_min_radius_scaled)
+
+    dilated_mask = dilate_big_areas(morphology_enhanced, min_cluster_area_scaled, kernel_size)
+
+    return dilated_mask
+
+
+def mark_small_areas(mask, max_hole_size, result_mask):
+    components, num_components = sp.ndimage.label(mask, np.ones((3, 3)))
+    slices = sp.ndimage.find_objects(components)
+    for label, slice in zip(range(1, num_components + 1),slices):
+        components_slice = components[slice] == label
+        if np.count_nonzero(components_slice) < max_hole_size:
+            result_mask[slice][components_slice] = True
+    return result_mask
+
+
+def dilate_big_areas(mask, min_area_size, dilate_radius):
+    components, num_components = sp.ndimage.label(mask, np.ones((3, 3)))
+    component = np.zeros(mask.shape, dtype=bool)
+    for label in range(1, num_components + 1):
+        np.equal(components, label, component)
+        if np.count_nonzero(component) > min_area_size:
+            tmp_mask = image_dilate(component, dilate_radius)
+            mask = mask | tmp_mask
+
+    return mask
+
+
+def fill_holes(mask, kernel_size, minimal_hole_size):
+    """
+    Fills holes in a given mask using iterative close + dilate morphological operations and filtering small patches.
+    @param mask: mask which holes are to be filled
+    @param kernel_size: size of the morphological element used to dilate/erode mask
+    @param minimal_hole_size: holes with area smaller than param are to be removed
+    """
+    nr = 1
+    morphology_element = get_circle_kernel(kernel_size)
+    while True:
+        new_mask = mask.copy()
+        # find connected components
+        components, num_components = sp.ndimage.label(np.logical_not(new_mask), np.ones((3, 3)))
+        slices = sp.ndimage.find_objects(components)
+        for label, slice in zip(range(1, num_components + 1), slices):
+            slice = extend_slices(slice,kernel_size+1000)
```

Comment:
I think that

    slice = extend_slices(slice, morphology_element.shape[0] * 2)

should be sufficient and should speed this up dramatically?